addons/sale_coupon/models/sale_coupon_program.py

```diff
@@ -237,7 +237,7 @@ def _filter_promo_programs_with_code(self, order):
         return self.filtered(lambda program: program.promo_code_usage == 'code_needed' and program.promo_code != order.promo_code)
 
     def _filter_unexpired_programs(self, order):
-        return self.filtered(lambda program: program.maximum_use_number == 0 or program.order_count < program.maximum_use_number)
+        return self.filtered(lambda program: program.maximum_use_number == 0 or program.order_count < program.maximum_use_number or (program.order_count == program.maximum_use_number and program == order.code_promo_program_id))
```

Comment:
Hm, correct me if I am wrong, but this is handling only the case of a promotion with a code, right?
The behavior here will remain unchanged for automatically applied promotion, but it actually isn't noticeable at a UX level since later in `recompute_coupon_lines()` there is a call to `_create_new_no_code_promo_reward_lines()` which will create the line juste after still removing it here.

As you prefer I guess, either:
1. Add a comment about that explaining that we only check `code_promo_program_id` and not `no_code_promo_program_ids` as it will be recreated later
2. Handle it the same way, something like 
```python
program in (order.code_promo_program_id + order.no_code_promo_program_ids)
```

I'd go for 2. so it is technically correct and we don't have to unlink/recreate line later for no reason

Reply:
as a side note, the lines becomes very long, `black` suggest to break it like this:

```py
        return self.filtered(
            lambda program: program.maximum_use_number == 0
            or program.order_count < program.maximum_use_number
            or (
                program.order_count == program.maximum_use_number
                and program == order.code_promo_program_id
            )
        )
```

Reply:
1. Why not using @nle-odoo suggestion? You did
```python
        return self.filtered(lambda program: program.maximum_use_number == 0
                             or program.order_count < program.maximum_use_number
                             or program
                             in (order.code_promo_program_id + order.no_code_promo_program_ids)
                             )
```
Which is obviously way less clear, losing lot of indent level for no reason IMHO

2. Regarding https://github.com/odoo/odoo/pull/83676#discussion_r797753957 I saw you did the change and requested a new review but.. Should I guess you investigate after my message and confirm what I said or just wrote the change ? It had to be confirmed, was just hinting you, but since you did not replied, just to be sure..

Reply:
> 1. Why not using @nle-odoo suggestion? You did
> 
> ```python
>         return self.filtered(lambda program: program.maximum_use_number == 0
>                              or program.order_count < program.maximum_use_number
>                              or program
>                              in (order.code_promo_program_id + order.no_code_promo_program_ids)
>                              )
> ```
> 
> Which is obviously way less clear, losing lot of indent level for no reason IMHO
> 
> 2. Regarding [[FIX] sale_coupon: last use of promotion program coupon fails #83676 (comment)](https://github.com/odoo/odoo/pull/83676#discussion_r797753957) I saw you did the change and requested a new review but.. Should I guess you investigate after my message and confirm what I said or just wrote the change ? It had to be confirmed, was just hinting you, but since you did not replied, just to be sure..

I did comment also asking about something but I figured out the answer that is why I deleted the comment

this is indeed only concerns programs with promo codes

Reply:
@rdeodoo  edited to adopt @nle-odoo suggestion. Thanks!